### ISSUE TITLE
IAM role credentials are not being renewed when using botocore>=1.25.0

### ISSUE THREAD

**Issue by aamattos** [boto/botocore#2670]
### Describe the bug

The `_evaluate_expiration(self, credentials)` method on utils.py is seting a `refresh_offset`  value that always range between 15mins and 25 mins. 
This method is updating the `expiry_time` data with the value `new_time = current_time + refresh_offset` 
The new value could end up being greater than the actual expiration time, leading to the token expiring before being renewed.

### Expected Behavior

IAM Role credentials being renewed before the expiration time

### Current Behavior

IAM Role credentials are not being renewed.

### Reproduction Steps

    def run(self):
        """main run method"""

        # Create an AWS client
        aws_config = Config(retries={"max_attempts": 10, "mode": "standard"})

        while True:

            s3_client = boto3.client(service_name="s3", region_name="us-east-1", config=aws_config)
            s3_client.list_objects_v2(Bucket="some_bucket")
            time.sleep(10)

### Possible Solution

The `_evaluate_expiration` method should check that the `new_time` value is always between `current_time` and `expiration`

### Additional Information/Context

I'm using a KIAM solution for authenticating via IAM Role.

### SDK version used

boto3==1.22.11

### Environment details (OS name and version, etc.)

Kubernetes 1.19, Python 3.7

**Comment by nateprewitt** [comment]
Hi @aamattos, how long are the credentials being provided through KIAM valid for and how are they being sourced?

The official IMDS interface provides credentials that are valid for 6 hours which are refreshed hourly. We're aware solutions like aws-vault, KIAM, kube2iam, etc. are utilizing this interface to inject alternative credentials as if they were emitted from IMDS, but those should still have a minimum validity of 15 minutes.

The new window provided in the code is technically 1-10 minutes due to refreshes being performed 15 minutes prior to expiration. Could you provide some more information about the credentials you're using to back KIAM and the exception you're hitting with the failure?

**Comment by aamattos** [comment]
> 

Hi @nateprewitt the solution we are using is: https://github.com/uswitch/kiam
We currently have a session duration of 30 mins.
I was able to debug the problem, and as you could see, before entering the `_evaluate_expiration` method, the expiration time is `9:14:20`. After the evaluation, the new expiration time is `9:19:07`. causing the application to fail at the original expiration time:

![Screenshot 2022-05-12 at 10 04 08](https://user-images.githubusercontent.com/1260299/168042172-30d041c8-fe19-4c95-932e-3bdb2f1a4668.png)


![Screenshot 2022-05-12 at 10 15 32](https://user-images.githubusercontent.com/1260299/168043035-9b70db6c-93eb-4a38-b9d9-f2e5ccc364c4.png)

**Comment by nateprewitt** [comment]
Hi @aamattos, that's correct. This is the intended functionality for the change in 1.25.0. Let me give a little more background about what you're seeing and then we can talk about solutions. I've left a TL;DR at the bottom for those just looking for a solution.

## Background

Instance Metadata Service (IMDS) is available for EC2 hosts to surface metadata, including credentials for roles attached to the machine. The IMDS instance provided by AWS will always provide credentials valid for 6 hours and, during normal operations, refresh them hourly.

During outage events, it's possible the local IMDS node for your instance may be unreachable for updates. That's where this change comes into place. In that case where credentials haven't been updated for 6 hours, the SDK will attempt to extend the credential validity beyond the published expiration time. This gives the option for the service authoring credentials to temporarily extend validity in exceptional situations, preventing wide-scale impact to SDK users.

## The problem

With actual IMDS, this extension works fine. It becomes problematic due to tools like KIAM that have been built to impersonate IMDS. They're inconsistent with what the service is intended to do, and between other implementations, so there's no standardized behavior. The authors of KIAM have moved away from using their tool because of these edge cases and have published a warning about it in the project README.

What we've done to minimize impact to these third-party solutions in ensure our refresh window is <= 10 minutes. This was chosen because STS (what KIAM uses to fetch credentials) cannot create credentials that are valid less than 15 minutes. We could be more aggressive with refreshes and check more often, but this brings a risk that during outage events, SDKs may start effectively DDoSing IMDS if all of them start trying to constantly refresh. We could also attempt a refresh when we reach the original expiration, but this is surprisingly complex with Boto's current credential providers. Accomplishing this would require a complete rewrite which has other impact implications for tools extending our credential interfaces.

The current 2-10 minute window works for most tools like KIAM because they're refreshed roughly every 5 minutes. It appears we have a gap specifically for KIAM though.

## Solutions

I'm taking a look at what impact would be if we shrunk the refresh window more. This may help, but is an imperfect solution because we're always going to have some level of race condition. There's also the possibility of allowing opt-out from this extension behavior, but we've generally avoided flags like this. It introduces potential inconsistency between environments which has often been difficult for customers to debug.

Looking at the [code for KIAM](https://github.com/uswitch/kiam/blob/80172c4e8d95bfdfd34188f6b44fedb7d170360e/cmd/kiam/server.go#L59), it performs refreshes by default 5 minutes before the published expiration time. This leaves a 5 minute gap where your credentials may not be refreshed in time. Setting this value to 10 minutes would mitigate impact in the short term. Long-term it may be worth considering moving off KIAM to more modern options like [IAM roles for Service Accounts](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html).

Let me know what your thoughts are on the above and if the refresh extension in KIAM is an acceptable option. Thanks!

## TL;DR

KIAM varies from how IMDS is supposed to function. Setting the [refresh window in KIAM to >=10m](https://github.com/uswitch/kiam/blob/80172c4e8d95bfdfd34188f6b44fedb7d170360e/cmd/kiam/server.go#L59) should mitigate the current impact. We're looking at potential options to fix on our side.

**Comment by aamattos** [comment]
Hi @nateprewitt thank you for the clear explanation.
We actually have plans to move to [IAM roles for Service Accounts](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html) as we know the KIAM solution is on maintenance mode, but this will take us some time and we probably will have both solutions co-existing for a while.
I will try to increase both, the `session-duration` and `session-refresh` flags on the KIAM server and see how it behaves.
I'll let you know about the outcomes.
Thanks again !
